### PR TITLE
[Snyk] Security upgrade transformers from 4.51.3 to 5.5.0

### PR DESCRIPTION
**Pull request by karly-quick**
snyk-top-banner

#### Snyk changed the following file(s):

- `tt-train/tools/weight-utils/requirements.txt`

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Learn about vulnerability in an interactive lesson of Snyk Learn.](https://learn.snyk.io/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":true,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"transformers","from":"4.51.3","to":"5.5.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-TRANSFORMERS-17675094"],"prId":"4efe68f2-fc5e-40dc-bccb-603d868297a8","prPublicId":"4efe68f2-fc5e-40dc-bccb-603d868297a8","packageManager":"pip","priorityScoreList":[180],"projectPublicId":"4aa8cc94-72b9-4901-bdae-903e8d6a0525","projectUrl":"https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-TRANSFORMERS-17675094"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by karly-quick** [comment]
Merge Risk: High

The upgrade from `transformers` v4 to v5 is a major release with significant and extensive breaking changes. This is a high-risk upgrade that will require substantial code modifications and testing.  
  
**Key Breaking Changes:**  
  
*   **PyTorch-Only Framework:** Support for TensorFlow (TF* classes) and JAX (Flax* classes) has been completely removed. The library is now PyTorch-only. Code using these backends will need a complete rewrite to PyTorch.  
*   **Environment Requirements:** `transformers` v5 requires **Python 3.10+** and **PyTorch 2.4+**.  
*   **Unified Tokenizer API:** The distinction between "slow" and "fast" tokenizers has been eliminated in favor of a single, unified tokenizer system. Imports and logic relying on separate tokenizer types will break.  
*   **Quantization API Overhaul:** Legacy arguments like `load_in_4bit` and `load_in_8bit` have been removed. A new, unified `quantization_config` must be used for loading quantized models.  
*   **Model Loading Behavior Change:** `from_pretrained()` now defaults to `dtype="auto"`, loading models in their native precision (e.g., float16) instead of float32. This can silently impact model outputs. To retain the old behavior, you must explicitly set `torch_dtype=torch.float32`.  
*   **Removed Features:** Several lesser-used features have been removed to simplify the library, including head masking, head pruning, and relative positional biases in some models.  
*   **API Parameter and CLI Changes:**  
    *   The `use_auth_token` parameter is deprecated and replaced by `token`.  
    *   The command-line tool `transformers-cli` is now `transformers`.  
  
**Recommendation:**  
This is a major migration effort. Developers must carefully review the official migration guide. Key actions include:  
1.  If using TensorFlow or JAX, you must migrate your entire implementation to PyTorch.  
2.  Ensure your environment meets the new Python 3.10+ and PyTorch 2.4+ requirements.  
3.  Update all tokenizer instantiation logic to use the new unified API.  
4.  Refactor all model loading calls to use the new `quantization_config` for quantized models and explicitly set `torch_dtype` if float32 precision is required.  
5.  Perform a find-and-replace for `use_auth_token` to `token`.  
  
Source: [GitHub Migration Guide](https://github.com/huggingface/transformers/blob/main/MIGRATION_GUIDE_V5.md), [Release Announcements](https://huggingface.co/blog/transformers-v5)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.

[//]: # (snyk:breakability:{"publicId":"e817c37e-b773-4b06-aabf-d617e1aff60c"})

**Comment by karly-quick** [comment]
### :warning: **Snyk checks are incomplete.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :warning: | Open Source Security| 0 | 0 | 0 | 0 | [ See details ](https://app.snyk.io/org/tt-test-rust/pr-checks/48e107c9-9bba-4342-9818-9e0a2b5115ec?refs=12f6398e7f22ca26a992d4010ac563ffea23f0b5&source=prChecksComment) |
| :warning: | Licenses| 0 | 0 | 0 | 0 | [ See details ](https://app.snyk.io/org/tt-test-rust/pr-checks/48e107c9-9bba-4342-9818-9e0a2b5115ec/license?refs=12f6398e7f22ca26a992d4010ac563ffea23f0b5&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/project/c45c259d-ddf0-4be1-960e-6f383f9dbef2/pr-check/070495b4-b429-4159-a9ce-dd07bda407ac?refs=12f6398e7f22ca26a992d4010ac563ffea23f0b5&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).
